### DISCUSSION TITLE
CentOS 7 build issues

### DISCUSSION THREAD

**Discussion by msercheli** [Autodesk/maya-usd#1473] [Q&A] [answered]
I am trying to build the maya-usd plugin in CentOS 7 using the build.py script provided, but "make" gets stuck (hanging forever) while building the "mayaUsdUI" target (at 88%):

```
...
Scanning dependencies of target _schemas
[ 87%] Building CXX object lib/usd/schemas/CMakeFiles/_schemas.dir/module.cpp.o
[ 87%] Building CXX object lib/usd/schemas/CMakeFiles/_schemas.dir/wrapTokens.cpp.o
[ 87%] Building CXX object lib/usd/schemas/CMakeFiles/_schemas.dir/wrapMayaReference.cpp.o
[ 87%] Building CXX object lib/usd/schemas/CMakeFiles/_schemas.dir/wrapALMayaReference.cpp.o
[ 88%] Linking CXX shared library _schemas.so
[ 88%] Built target _schemas
Scanning dependencies of target mayaUsdUI_autogen
[ 88%] Automatic MOC and UIC for target mayaUsdUI
```

Using:
- maya-usd (tag) v0.9.0
- make 3.82
- gcc 6.3.0
- Maya 2022.0 + devkit
- qtbase 5.15.2
- USD 0.20.8 (20.08)
- Python 2.7

CMakeLists.txt options:

- BUILD_ADSK_PLUGIN ON
- BUILD_PXR_PLUGIN OFF
- BUILD_AL_PLUGIN OFF
- BUILD_TESTS OFF

The build.py script was able to find Python, USD, UFE (2.0.0) and Qt.
Running the build.py script with "--verbosity" set to 2 doesn't give me any extra info.

Seems like "make" is hanging during the MOC execution due to the CMake directive `set(CMAKE_AUTOMOC ON)` at `lib/usd/ui/CMakeLists.txt` line 19.

I verified that the executable "moc" file (from qtbase build) is in the PATH and available.
Any help is much appreciated.
Thanks in advance.

Marcelo

**Comment by seando-adsk** [comment]
Just to verify when calling build.py are you passing the flag?
`  --qt-location QT_LOCATION
                        Directory where Qt is installed.`

You could try adding this:
` --build-args="-DQT_VERSION=5.15"`

I see [here](https://github.com/Autodesk/maya-usd/blob/dev/CMakeLists.txt#L137)  that we made an assumption and set qt version to 5.12, but for 2022 it should be 5.15.

**Reply by msercheli** [reply]
I am passing the `--qt-location`, but I am not passing `-DQT_VERSION=5.15`.
Looking into the build directory, under `lib/usd/ui/mayaUsdUI_autogen/5WMDHDZAMM` I can find 2 `.cpp` files:
- moc_mayaLayerEditorWindow.cpp
- moc_stageSelectorWidget.cpp

So, it seems the build script is finding my Qt5 and the executable `moc` as well.
But I can try specifying the qt version like you suggested to see if it makes any difference.

In the CMakeLists ([https://github.com/Autodesk/maya-usd/blob/dev/CMakeLists.txt#L137](url)) that you mentioned, is it trying to find 5.12 as minimum version, but if it finds 5.15 it should be fine?

**Reply by msercheli** [reply]
It is worth mention that I am trying to build it against a `qtbase` build (5.15.2). That is the `qt location` I am passing to the build system.

Although the devkit provides Qt5, I tried to give the devkit location as the qt location, but it could not find the `bin` directory for Qt since it is a subdirectory of `devkit`. for that reason, I built qtbase myself and I am using it.

I know Maya needs its own Qt5 to run. Is it the same for the MayaUsd plugin? AM I forced to use the Qt5 provided by the devkit to build the USD plugin as well?

**Reply by msercheli** [reply]
Sean,

I tried providing the QT_VERSION but no luck. Same thing. "make" still hangs at 88% while MOC is executing.

**Comment by HamedSabri-adsk** [comment]
@msercheli  No you are not forced to use the devkit to build the UI parts of MayaUSD plugin. 

I can completely understand why you couldn't point `--qt-location ` to `devkit` since directory structure there is not well organized. 

I build qt from source myself and use it. One could also use the  binaries https://www.qt.io/download

I use mixed versions of Qt ( Qt5.9 , 5.12, Qt 5.15 ) on different machines and they all work fine.

**Reply by msercheli** [reply]
Hi Hamed,

You mentioned that you are building the plugin just fine against Qt on different machines. Are you building specifically on CentOS 7 (default installation)? Or are you building using some docker container that is providing a different environment?
If you could share your environment, than I can compare with mine.

**Reply by HamedSabri-adsk** [reply]
Hi Marcelo,

I don't use docker. On my Linux box, I am using `Qt5.12.5` pre-compiled binaries https://www.qt.io/download which --qt-location points at.

`python build.py --maya-location "/home/sabrih/maya/runtime" --pxrusd-location "/home/sabrih/Usd2102" --devkit-location "/home/sabrih/devkitBase" --stages=configure,build,install  --generator Ninja --build-args=-DBUILD_WITH_PYTHON_3=ON,-DCMAKE_MAKE_PROGRAM="/home/sabrih/ninja" --qt-location /home/sabrih/Qt/5.12.5/gcc_64 /home/sabrih/workspace`

On MacOS, I have `5.9` that I built from source while ago. On Windows, I am using `Qt5.15` pre-compiled binaries.

I also just built the project with `make` on Centos7 and don't see any hang. I usually throw all my cores when building the project. 

```
    Building with settings:
      Source directory          /home/sabrih/Desktop/maya-usd
      Workspace directory       /home/sabrih/Desktop/workspace
      Build directory           /home/sabrih/Desktop/workspace/build/RelWithDebInfo
      Install directory         /home/sabrih/Desktop/workspace/install/RelWithDebInfo
      Variant                   RelWithDebInfo
      Python Debug              Off
      CMake generator           Default
      Build Log                 /home/sabrih/Desktop/workspace/build/RelWithDebInfo/build_log.txt
      Build arguments           ['-DBUILD_WITH_PYTHON_3=ON', '']
      Stages arguments          ['configure', 'build', 'install']
Success MayaUSD build and install !!!!
Elapsed time: 00:04:27
```
Can you please share your system secs?

**Reply by msercheli** [reply]
Linux version: CentOS-7.9.2009
Kernel version: 3.10.0-1160.25.1.el7.x86_64
Make version: GNU Make 3.82
glibc version (extracted using ldd --version): ldd (GNU libc) 2.17
ninja version; 1.10.2

Here is a full list of all the packages involved in my build and their versions. Most of them are USD dependencies:
```
Jinja2-2.11.3
MarkupSafe-1.1.1
OpenSubdiv-3.4.3
PyOpenGL-3.1.5
PySide2-5.15.2
PyYAML-5.4.1
blosc-1.5.0
boost-1.70.0
cmake-3.14.6
gcc-6.3.0
glew-2.0.0
glfw-3.4.0
libjpeg-9.2
libpng-1.6.37
libtiff-4.0.7
maya-2022.0
maya_devkit-2022.0
ocio-1.0.9
oiio-2.1.16.0
openexr-2.2.0
openvdb-6.1.0
osl-1.9.13
ptex-2.1.28
python-2.7.5
qtbase-5.15.2
shiboken2-5.15.2
tbb-2017.6
usd-20.08
```

My setup is a little special. I am using REZ [https://github.com/nerdvegas/rez](url) to control my build environment.
All the packages you see in the above list were build using REZ system.
But aside from REZ, the final build environment has all those packages.

**Reply by HamedSabri-adsk** [reply]
To me it looks like you have all the right packages. I am not too familiar with REZ but recall some people reported issues with it in the past. I am not entirely sure if it is related to you situation.

**Comment by HamedSabri-adsk** [comment]
@msercheli  Can you try `Ninja` instead of `make ` and see if it fixes the issue for you?

**Reply by msercheli** [reply]
Hi Hamed,

I tried ninja before as well. Same result.

Unfortunately, I am forced to use -j1 because every time I try running the build with --jobs > 1, at some point (random point) it hangs. I learned online that it happens in many other projects. Some claim to be related to CPU heating issues and others claim to be related to racing conditions on make targets. All that to say that my build attempts are slow.

A friend of mine suggested hooking up a "process monitor" to see what is stuck. I will research about it to see how I can do.

**Comment by msercheli** [comment]
I am building the plugin using GCC 6.3.0 since the toolset-6 is not available anymore (it provides GCC 6.3.1). But from what I read online, 6.3.1 was merely a snapshot at some point in time during the development of GCC. The GCC repo makes available only 6.3.0, jumping next to 6.4.0 and so on. Apparently, GCC 6.3.0 that is available in the repo is actually the latest of the 6.3 series.

But is it possible that the GCC 6.3.0 is causing the issue? I am also using CMake 6.14.6, which I believe is a valid version for the MayaUsd plugin.

**Reply by HamedSabri-adsk** [reply]
It could be it but I am not 100% sure. I don't think any of us has used "6.3.0" before.  I have both gcc toolset 6.3.1 and 9.3.1 and switch between them without any issues. It would be worth trying those versions and see if it helps.  

Also, the CMake version that you mentioned cant be valid. CMake major version starts at 3 ( e.g 3.19.4 ).  Any CMake version after 3.13 should be fine.

https://github.com/Autodesk/maya-usd/blob/dev/doc/build.md

**Reply by msercheli** [reply]
Apologies. CMake version was a typo. I meant 3.14.6

Forgive my limited knowledge here, but how do you install GCC 6.3.1 in CentOS 7?
Everywhere I read, it says to install toolset-6, which is not available anymore.
I am following these steps:

```
$ sudo yum install centos-release-scl
$ sudo yum-config-manager --enable rhel-server-rhscl-7-rpms
$ sudo yum install devtoolset-6
```
The last command returns:
`No package devtoolset-6 available.`

I built from source, but there is no 6.3.1 tag in [https://github.com/gcc-mirror/gcc](url) or [https://ftp.gnu.org/gnu/gcc/](url)
I could only find 6.3.0

**Comment by HamedSabri-adsk** [comment, marked as answer]
No worries. Here is the instruction to install devtoolset-6 on Centos from the vault:

```
devtoolset-6, which includes GCC 6 on CentOS has been retired from the main repo and moved into the vault. To install:

# download the packages, install may fail with "no public key found"
sudo yum-config-manager --add-repo=http://vault.centos.org/7.6.1810/sclo/x86_64/rh/

# to fix "no public key found"
cd /etc/pki/rpm-gpg
ls # confirm RPM-GPG-KEY-CentOS-SIG-SCLo exists
sudo rpm --import RPM-GPG-KEY-CentOS-SIG-SCLo
rpm -qa gpg* # confirm key with substring f2ee9d55 exists

# to install devtoolset-6
sudo yum install devtoolset-6

# disable the vault after successful install
sudo yum-config-manager --disable vault.centos.org_7.6.1810_sclo_x86_64_rh_
```

**Reply by msercheli** [reply]
That is awesome.
Thank you very much. Installed like a charm.

I will re-build USD, Qt5 and eventually the MayaUsd plugin and will report back here. Hopefully it will work fine.

**Reply by msercheli** [reply]
First, many thanks to Hamed and Sean for the help.

The original issue related to CMAKE_AUTOMOC, where the build would just freeze/hang indefinitely, is being caused by an issue in my machine (CentOS 7).

Unfortunately, for some reason that I cannot find out, I cannot build using `make --jobs` larger than 1. Any number larger than 1 will cause my build to hang at some point (random) of the build. This is happening in any build, not just the maya-usd plugin.

The solution for me is to specify `--jobs 1` while building. Although, currently, the `CMAKE_AUTOMOC` is not respecting the `--jobs` argument passed to the build system.

I would recommend modifying the `CMakeLists.txt` in the build root to respect the jobs argument by setting the CMake directive `CMAKE_AUTOGEN_PARALLEL` to the number passed by the `--jobs` argument.

My issue is solved for now since I passed `-DCMAKE_AUTOGEN_PARALLEL=1`.

I will flag this discussion as `answered` and any new issues I will open a new one.

**Reply by HamedSabri-adsk** [reply]
@msercheli You are welcome. I have never encountered such behavior before. I did a google search to see if someone else has reported such issue with CMAKE_AUTOMOC but my search came up empty.  

One thing you could try is to get a more recent version of CMake ( e.g 3.19.x ) and see if it helps. 

As for `CMAKE_AUTOGEN_PARALLEL`, I didn't know such flag exist. I just read the docs and it seems this option was added in 3.11. 

https://cmake.org/cmake/help/latest/variable/CMAKE_AUTOGEN_PARALLEL.html

I don't see any issues with having this flag enabled and would be nice to have it ON. I will have a PR to add it in.

**Reply by msercheli** [reply]
Hi Hamed,

I believe the default behavior of the `CMAKE_AUTOGEN_PARALLEL` flag is to use as many cores as available for the MOC processing. Only if set it locks to a specific number. So, as it is not, the maya-usd plugin is processing MOC in parallel. But it is not respecting the --jobs passed. Maybe the PR can address that. I will leave it up to you. :-)
